server/src/main/java/org/elasticsearch/action/support/RefCountingRunnable.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.util.concurrent.CountDown;
+import org.elasticsearch.core.AbstractRefCounted;
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.Releasable;
+
+/**
+ * A mechanism to trigger an action on the completion of some (dynamic) collection of other actions. Basic usage is as follows:
+ *
+ * <pre>
+ * try (var refs = new RefCountingRunnable(finalRunnable)) {
+ *     for (var item : collection) {
+ *         runAsyncAction(item, refs.acquire()); // releases the acquired ref on completion
+ *     }
+ * }
+ * </pre>
+ *
+ * The delegate action is completed when execution leaves the try-with-resources block and every acquired reference is released. Unlike a
+ * {@link CountDown} there is no need to declare the number of subsidiary actions up front (refs can be acquired dynamically as needed) nor
+ * does the caller need to check for completion each time a reference is released. Moreover even outside the try-with-resources block you
+ * can continue to acquire additional listeners, even in a separate thread, as long as there's at least one listener outstanding:
```

Comment:
Isn't this really risky - unless you're very careful, the async actions could complete really quickly, leading to the runnable running before you're ready.

Reply:
I'm not sure I understand. Callers can avoid the runnable running before they're ready by keeping hold of one or more references.

Reply:
Indeed this is the whole point of this bit of the example code below:

```
 *         var itemRef = refs.acquire(); // delays completion while the background action is pending
  *         executorService.execute(() -> {
  *             try (var ignored = itemRef) {
  *                 if (condition(item)) {
  *                     runOtherAsyncAction(item, refs.acquire());
  *                 }
  *             }
  *         });
```

I'm not seeing the risk of an early trigger here - indeed an early trigger would result in assertions tripping.

Reply:
Ah yes, as long as there's one listener outstanding. OK.